Implement hns validation webhook

yiqigao217 opened the pull request:
Add HierarchicalNamespace and Namespace webhooks. They will allow all
requests from HNC SA.

HNS webhook will deny:
1) creating hns in a forbidden (excluded) namespace;
2) creating hns with an existing namespace name;
3) deleting hns if the owned namespace doesn't allow cascading deletion.

NS webhook will deny:
1) deleting ns if it's owned, no matter if it allows cascading deletion,
since users should delete the hns instance in the owner instead;
2) deleting ns if it has owned ns and it doesn't allow cascading
deletion.

Tested with new unit tests and manually tested on GKE cluster to see if
it works as expected and no conflict with cascading deletion implemented
earlier.

Fixes #458 
Part of #457

yiqigao217 commented:
/assign @adrianludwin 
/assign @sophieliu15 
cc @adrianchung

sophieliu15 commented on incubator/hnc/pkg/validators/hierarchical_namespace.go:
```diff
@@ -0,0 +1,122 @@
+
+	if req.op == v1beta1.Create {
+		if reconcilers.EX[pnm] {
+			msg := fmt.Sprintf("The namespace %s is not allowed to create subnamespaces. Please create hns in a different namespace.", pnm)
```
Maybe use the full name instead of acronym? Or jut use "subnamespaces" as the first sentance.

sophieliu15 commented on incubator/hnc/pkg/validators/hierarchical_namespace.go:
```diff
@@ -0,0 +1,122 @@
+
+		if cns.Exists() {
+			msg := fmt.Sprintf("The requested namespace %s already exists. Please use a different name.", cnm)
+			log.V(1).Info(msg)
```
Do we need to log the message here? IIUC, it will be logged in line 56?

adrianchung commented on incubator/hnc/pkg/validators/hierarchical_namespace.go:
```diff
@@ -0,0 +1,122 @@
+	}, nil
+}
+
+func (v *HierarchicalNamespace) InjectDecoder(d *admission.Decoder) error {
```
Is this function used in this PR? I'd expect to see a test for something like this, but don't see anything associated with it -- either that or as a part of an interface definition. Am I missing some understanding here?

adrianludwin replied:
You mean `InjectDecoder`? It's part of the boilerplate code that's called by `controller-runtime`, I'm pretty sure nothing will compile if it's not there.

yiqigao217 replied:
The documentation of the interface DecoderInjector is here: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/webhook/admission#DecoderInjector.

adrianchung replied:
I tried just deleting `InjectDecoder` and `InjectClient` methods out of .pkg/validators/hierarchy.go as a test and the binary still compiles and deploys. I'm not one to argue with the compiler. As an experiment can you see what'd happen if we took this out and see if there's any side effects? I don't want to block the review or anything -- you're just likely closest to trying this out!

adrianludwin replied:
Hmm. Are you sure the webhook is running correctly after this? Without `InjectDecoder`, I'm not sure how the `decoder` field is set, which means it should cause a nil pointer exception the first time we try to decode something. Perhaps it just looks fine because you never created a HierarchicalNamespace object, but as soon as you do, HNC will crash?

I would have _thought_ that `InjectDecoder` was checked at compile time, but perhaps it's only checked at runtime with a [type switch](https://tour.golang.org/methods/16).

yiqigao217 replied:
Thanks for bringing this up! I am also curious so I tested with removing the `InjectDecoder` in namespace webhook and `make deploy`. I got errors when trying to delete a namespace. The errors was the same as when I had dereference of an nil pointer similar to #603 .

adrianludwin replied:
Why didn't you need the decoder when trying to _create_ a namespace?

yiqigao217 replied:
I was just testing if it would work if the `InjectDecoder` is removed from the code. And it turned out it will fail, so we need it.

adrianludwin replied:
👍

adrianludwin commented on incubator/hnc/pkg/validators/namespace.go:
```diff
@@ -0,0 +1,121 @@
+	v.Forest.Lock()
+	defer v.Forest.Unlock()
+
+	ns := v.Forest.Get(req.ns.Name)
+
+	// Check whether the deleting namespace has owned namespaces if it doesn't allow cascading deletion.
+	if req.op == v1beta1.Delete && !ns.AllowsCascadingDelete() {
```
This might be a bit easier to read if you had more early exits:

```
if req.op != v1beta1.Delete {
  return allow("") // we only care about deletions
}

if ns.AllowsCascadingDelete {
  return allow("") // go ahead, make my day
}

// Check to see if there are any HNSes that can't be deleted
...
```
```

yiqigao217 replied:
Done, I put first two `if`s into one.

adrianludwin commented on incubator/hnc/pkg/validators/namespace.go:
```diff
@@ -0,0 +1,121 @@
+		hnses := []string{}
+		for _, cnm := range ns.ChildNames() {
+			cns := v.Forest.Get(cnm)
+			if cns.IsOwned {
+				hnses = append(hnses, cnm)
+			}
```
Shouldn't this be `if cns.IsOwned && !cns.AllowsCascadingDelete()`? If so, maybe change `hnses` to `cantDelete` or something like that, to make it clear why things get added to that list.

If I'm correct, and this is a bug, maybe add a unit test for this too?

yiqigao217 replied:
Yes good catch it's a bug. Say the owner doesn't allowCD but all its owned NS allows, then it should return `allow()`. Good catch on this corner case, will add the unit test on that and manual test it too.

yiqigao217 replied:
Added new unit tests and manually tested with setting `allowCascadingDelete`s (the webhook and cascading deletion all worked as expected).

adrianludwin commented on incubator/hnc/pkg/validators/namespace.go:
```diff
@@ -0,0 +1,121 @@
+			}
+		}
+		if len(hnses) != 0 {
+			msg := fmt.Sprintf("The namespace has owned namespace(s): %s. Please set allowCascadingDelete flag to true first.", hnses)
```
Maybe `Please set allowCascadingDelete first either in the parent namespace or in all the owned children`?

yiqigao217 replied:
Done, I also added a list of owned namespaces with `allowCascadingDelete` unset.

yiqigao217 commented on incubator/hnc/pkg/validators/hierarchical_namespace_test.go:
```diff
@@ -0,0 +1,63 @@
+		fail bool
+	}{
+		{name: "ok-create", op: v1beta1.Create, pnm: "foo", cnm: "brumpf"},
+		{name: "ok-delete", op: v1beta1.Delete, pnm: "foo", cnm: "baz"},
```
FYI updated in line 21 by `createOwnedNamespace()`.

jiachengxu commented on incubator/hnc/pkg/validators/hierarchical_namespace.go:
```diff
@@ -0,0 +1,119 @@
+// Note: the validating webhook FAILS CLOSE. This means that if the webhook goes down, all further
+// changes are forbidden.
+//
+// +kubebuilder:webhook:path=/validate-hnc-x-k8s-io-v1alpha1-hierarchicalnamespaces,mutating=false,failurePolicy=fail,groups="hnc.x-k8s.io",resources=hierarchicalnamespaces,verbs=create;update;delete,versions=v1alpha1,name=hierarchicalnamespaces.hnc.x-k8s.io
```
why the `update` is needed? In your `Handle` method, it only handles `create` and `delete` operations.

adrianludwin replied:
Nice catch! We should probably reject any updates to HNSes (it doesn't have any real fields anyway), or else take `update` out of this list.

yiqigao217 replied:
Good catch, thanks! Removed.

yiqigao217 replied:
I manually tested with `update` removed. Seems to reject all `update` requests, we need to have it listed here and in `Handle` reject all `update` requests.

adrianludwin replied:
How did you try to update it? You mean with labels and stuff? Yeah, perhaps we should allow those.

I'm surprised that an action that we _didn't_ list as part of the validator would be blocked. What was the error message with `update` removed from the webhook config?

adrianludwin replied:
Note that `update` is still present here.

yiqigao217 replied:
> How did you try to update it? You mean with labels and stuff? Yeah, perhaps we should allow those.

I used `kubectl edit hns child -n parent`
> I'm surprised that an action that we _didn't_ list as part of the validator would be blocked. What was the error message with `update` removed from the webhook config?

The action not listed will be `allowed`, not blocked. When `update` removed, I could successfully edit the hns instance.

adrianludwin replied:
Ah, I see. I think we should allow them to let labels and annotations be added, wdyt?

yiqigao217 replied:
But that's by HNC SA, right? Not by users? We only deny `update` if it's not from SA.

adrianludwin replied:
No, if a user wants to go add a couple of labels to the HNS objects (_not_ the namespaces), I don't see why we'd stop them.

yiqigao217 replied:
Got you. Sure, then I just need to remove the `update` verb and clean it up in `handle`.

yiqigao217 commented:
PTAL, I updated the namespace webhook to check if all owned namespaces allowCD and added the new unit tests. I also updated the HNS webhook to deny all `update` requests and added one unit test case. @adrianludwin , @jiachengxu

adrianludwin commented on incubator/hnc/pkg/validators/hierarchical_namespace.go:
```diff
@@ -0,0 +1,123 @@
+	cnm := req.hns.Name
+	cns := v.Forest.Get(cnm)
+
+	if req.op == v1beta1.Update {
```
Maybe restructure this as a switch statement, with any unexpected op resulting in `log.Error`? We should let unexpected ops through though, since it's probably just a webhook misconfiguration.

I'm thinking we _should_ allow `Update` though in case people want to put labels or annotations onto them. You can't change the namespace or name, which are the only properties we care about anyway.

yiqigao217 replied:
As discussed above, i will remove all `update` related code since we will allow `update` requests.

yiqigao217 commented:
`update` verb is removed in HNS webhook and the `update` request handling and the unit test are all removed. @adrianludwin

adrianludwin commented:
/lgtm
/approve
/hold
/assign @rjbez17
